indy_node/test/txn_author_agreement/test_recover_taa_from_ledger.py

```diff
@@ -0,0 +1,104 @@
+    rest_pool = txnPoolNodeSet[:-1]
+    disconnect_node_and_ensure_disconnected(looper,
+                                            txnPoolNodeSet,
+                                            node_to_stop.name,
+                                            stopNode=True)
+    looper.removeProdable(name=node_to_stop.name)
+
+    # Step 2. Path all the rest nodes for using old version TAA handler
+
+    # it's ugly but it works
+    globals()['CURRENT_TXN_VERSIONS'][TXN_AUTHOR_AGREEMENT] = '1'
+    for node in rest_pool:
+        handler = node.write_manager.request_handlers.get(TXN_AUTHOR_AGREEMENT)[0]
+        orig_handlers[node.name] = handler
+        handler_for_v_1 = node.write_manager._request_handlers_with_version.get((TXN_AUTHOR_AGREEMENT, "1"))[0]
+        node.write_manager.request_handlers[TXN_AUTHOR_AGREEMENT] = [handler_for_v_1]
+
+    # Step 3. Send TAA txn in old way
+    text = randomString(1024)
+    version_1 = randomString(16)
+
+    sdk_send_txn_author_agreement(looper, sdk_pool_handle, sdk_wallet_trustee, text, version_1)
+
+    # Step 4. return original TAA handlers back
+
+    # it's ugly but it works
+    globals()['CURRENT_TXN_VERSIONS'][TXN_AUTHOR_AGREEMENT] = '2'
+    for node in rest_pool:
+        node.write_manager.request_handlers[TXN_AUTHOR_AGREEMENT] = [orig_handlers[node.name]]
+
+    # Step 5. Send another TAA txn in new way without optional parameters
+    text_2 = randomString(1024)
+    version_2 = randomString(16)
+    sdk_send_txn_author_agreement(looper, sdk_pool_handle, sdk_wallet_trustee, text_2, version_2)
+
+    # Step 6. Send another TAA txn in new way without optional parameter
+    text = randomString(1024)
+    version_3 = randomString(16)
+    sdk_send_txn_author_agreement(looper, sdk_pool_handle, sdk_wallet_trustee, text, version_3)
+
+    # Step 7. Send taa updating for the second taa transaction (for checking txn with optional parameter)
+    retired_time = int(time.time()) + 20
+    sdk_send_txn_author_agreement(looper, sdk_pool_handle, sdk_wallet_trustee, text_2, version_2, retired=retired_time)
+
+    # Step 8. Ensure, that all TAAs was written
+    res_1 = sdk_get_txn_author_agreement(looper, sdk_pool_handle, sdk_wallet_trustee, version=version_1)[1]
+    assert TXN_AUTHOR_AGREEMENT_DIGEST not in res_1['result']['data']
+    assert TXN_AUTHOR_AGREEMENT_RATIFIED not in res_1['result']['data']
+    assert res_1['result']['data'][TXN_AUTHOR_AGREEMENT_VERSION] == version_1
+
+    res_2 = sdk_get_txn_author_agreement(looper, sdk_pool_handle, sdk_wallet_trustee, version=version_2)[1]
+
+    assert TXN_AUTHOR_AGREEMENT_DIGEST in res_2['result']['data']
+    assert TXN_AUTHOR_AGREEMENT_RATIFIED in res_2['result']['data']
+    assert res_2['result']['data'][TXN_AUTHOR_AGREEMENT_VERSION] == version_2
+    assert res_2['result']['data'][TXN_AUTHOR_AGREEMENT_RETIRED] == retired_time
+    assert res_2['result']['data'][TXN_AUTHOR_AGREEMENT_RATIFIED] == retired_time - 20
```

Comment:
This line looks incorrect. Currently ratified time is a time when txn was ordered (pp_time), and we can not predict it.
I think the following check is enough:
 ```
 assert res_2['result']['data'][TXN_AUTHOR_AGREEMENT_RATIFIED] is not None
 assert res_2['result']['data'][TXN_AUTHOR_AGREEMENT_RATIFIED] < retired_time 
```